colossus/src/main/scala/colossus/protocols/memcache/Memcache.scala

```diff
@@ -10,39 +10,149 @@ import java.util.zip.{Deflater, Inflater}
+ *
+ * We also make sure that the keys are "well formed", removing all control characters and spaces, which
+ * are not allowed by Memcached
+ *
+ * I've grabbed a snippet from memcached docs about ttls found here: https://github.com/memcached/memcached/blob/master/doc/protocol.txt
+ *
+ * TTLs can be a unix timestamp for an integer that is no more than 30 days in seconds (2592000 seconds). If the TTL is greater
+ * than 30 days in seconds, the server will consider it to be real Unix time value rather than an offset from current time. 
+ * 
+ */
+
 
 object MemcacheCommand {
   val RN = ByteString("\r\n")
+  val SP = ByteString(" ")
+
+  private def FormatCommand(command: ByteString, key: ByteString, data: Option[ByteString] = None, ttl: Option[ByteString] = None, flags: Option[ByteString] = None): ByteString = {
+    val b = new ByteStringBuilder
+    b.append(command)
+    b.append(SP)
+    
+    val k = key.filter(_ > ' ').slice(0, 250) // Strip all control characters and space from the key, limit to 250
+    b.append(k)
+    
+    flags.foreach { f =>
+      b.append(SP)
+      b.append(f)
+    }
+    
+    ttl.foreach { t =>
+      b.append(SP)
+      b.append(t)
+    }
+
+    data.foreach { d  =>
+      b.append(SP)
+      b.append(ByteString(s"${d.size}"))
+      b.append(RN)
+      b.append(d)
+    }
+
+    b.append(RN) 
+    b.result
+  }
 
   case class Get(key: ByteString) extends MemcacheCommand {
-    def bytes (compressor: Compressor)= ByteString("get ") ++ key ++ RN
+    def bytes (compressor: Compressor)= FormatCommand(ByteString("get"), key)
   }
   object Get {
     def apply(key: String): Get = Get(ByteString(key))
   }
-  case class Set(key: ByteString,  value: ByteString) extends MemcacheCommand {
-
+  case class Set(key: ByteString,  value: ByteString, ttl: Integer) extends MemcacheCommand {
+  
     def bytes(compressor: Compressor) = {
-      val b = new ByteStringBuilder
       val data = compressor.compress(value)
-      b.append(ByteString("set ") )
-      b.append(key )
-      b.append(ByteString(s" 0 0 ${data.size}") )
-      b.append(RN )
-      b.append(data)
-      b.append(RN)
-      b.result
+      FormatCommand(ByteString("set"), key, Some(data), Some(ByteString(s"${ttl}")), Some(ByteString(s"0")))
     }
   }
   object Set {
-    def apply(key: String, value: String): Set = Set(ByteString(key), ByteString(value))
+    def apply(key: String, value: String, ttl: Integer = 0): Set = Set(ByteString(key), ByteString(value), ttl)
+  }
+  object Add {
+    def apply(key: String, value: String, ttl: Integer = 0): Add = Add(ByteString(key), ByteString(value), ttl)
+  }
+  case class Add(key: ByteString,  value: ByteString, ttl: Integer) extends MemcacheCommand {
+  
+    def bytes(compressor: Compressor) = {
+      val data = compressor.compress(value)
+      FormatCommand(ByteString("add"), key, Some(value), Some(ByteString(s"${ttl}")), Some(ByteString(s"0")))
```

Comment:
so it looks to me like every time you call FormatCommand, you're converting everything into ByteStrings.  It might make more sense to have the method simply take strings/integers as params and do the conversions itself.  Also it might be a good idea to put all the strings into constants inside an object.  Neither of these are very important here since all the code is self-contained, but they're things we try to do in general.